src/paths/steiner.cpp

```diff
@@ -0,0 +1,586 @@
+/*
+   IGraph library.
+   Copyright (C) 2022  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+#include "igraph_paths.h"
+
+#include "igraph_components.h"
+#include "igraph_interface.h"
+#include "igraph_matrix.h"
+#include "igraph_structural.h"
+
+#include "core/exceptions.h"
+#include "core/interruption.h"
+
+#include <cmath>
+#include <algorithm>
+#include <iterator>
+#include <map>
+#include <set>
+#include <vector>
+
+
+typedef std::set<igraph_integer_t> int_set;
+typedef std::map<std::set<igraph_integer_t>, igraph_integer_t> dictionary;
+
+/*
+ *  Generates Subsets of length > 1 for list of integer values.
+ *  As algorithm computes the minimum distance between set of variables ranging from
+ *  length 2 to steiner_terminals - 1, we need these subsets.
+ *  for e.g [2,3,4] --> [ [2,3], [2,4], [3,4], [2,3,4] ]
+ *  Populates the subsetMap data structure as well.
+ *
+ */
+static igraph_error_t generateSubsets(const igraph_vector_int_t *steinerTerminals, igraph_integer_t n, igraph_integer_t graphsize, dictionary& subsetMap, std::set<int_set> & allSubsets) {
+    if (n > sizeof(igraph_integer_t)){
+        IGRAPH_ERROR("igraph_integer_overflow detected. The given number of terminals is more than what the computer can handle.",IGRAPH_EINVAL);
+    }
+    igraph_integer_t count = ((igraph_integer_t)1 << n);
+    igraph_integer_t subsetIndex = graphsize;
+
+    // The outer for loop will run 2^n times to get all subset .
+    // Here variable i will act as a binary counter
+
+    for (igraph_integer_t i = 2; i < count; i++) {
```

Comment:
It's probably nice to add a comment on why this can start at 2.

Reply:
Yes. I added a comment saying that algorithms needs subsets of at least size 2 as distance for singleton can be easily calculated using adjacency matrix